### DISCUSSION TITLE
Is file ordering from globs supposed to be consistent?

### DISCUSSION THREAD

**Discussion by deviantintegral** [gulpjs/gulp#2747] [Help]
We ran into a bug today that appears to be caused by how gulp returns files from `src()`. Every time it runs, the order of files returned are different, leading to inconsistent builds.

I created a minimal example to show this that globs all JS files in the `node_modules` directory, hashes the returned paths, and then does the same after sorting. We can see each time the hashes of the original set of files change:

https://github.com/deviantintegral/gulp-sorting-example

```console
$ for i in {1..10}; do node index.js; done;
Unsorted hash is: 3aa0c0464a44b616dde4d645c91e44f8
Sorted hash is: abad07ce762de2827d06d5e182e95171
Unsorted hash is: 1a47c37b5f63e21d7a7cbb674cdc587d
Sorted hash is: abad07ce762de2827d06d5e182e95171
Unsorted hash is: 1f7f472a6542bc1a44c486ec8c906c64
Sorted hash is: abad07ce762de2827d06d5e182e95171
Unsorted hash is: 2be2814f2ae8d98a357f969e7bbce0f9
Sorted hash is: abad07ce762de2827d06d5e182e95171
Unsorted hash is: 0a770d1d0fe340028ee877a20cb8e91b
Sorted hash is: abad07ce762de2827d06d5e182e95171
Unsorted hash is: e59f8a0e5abc420bb502deca1e82274c
Sorted hash is: abad07ce762de2827d06d5e182e95171
Unsorted hash is: d1fe01c7fa4c987f1ec3b5aa3229dd3e
Sorted hash is: abad07ce762de2827d06d5e182e95171
Unsorted hash is: 96b1e6a53e5375e8bdb4f2fc6e771aca
Sorted hash is: abad07ce762de2827d06d5e182e95171
Unsorted hash is: 7ee390da2f1cf5da02803d628ebcb396
Sorted hash is: abad07ce762de2827d06d5e182e95171
Unsorted hash is: 08669e98fcbe3a6970faca7c01ab5496
Sorted hash is: abad07ce762de2827d06d5e182e95171
```

As I found, there is a plugin to sort files, but I was surprised to see this not be used by default everywhere in various projects. As well, this seems to break expectations for users from other ecosystems. For example, we can see bash globs are both consistent and sorted:

```console
acutus:gulp/ andrew$ ls node_modules/**/*.js | md5sum
226330bcb8dd4cc106a9a2f2818e16cd  -
acutus:gulp/ andrew$ ls node_modules/**/*.js | sort | md5sum
226330bcb8dd4cc106a9a2f2818e16cd  -
```

**Comment by bronzehedwick** [comment]
Ooph, yeah, this is an unpleasant surprise. I would expect glob file ordering to be consistent.
